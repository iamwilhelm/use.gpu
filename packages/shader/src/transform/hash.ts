export const C1 = 0xcc9e2d51;
export const C2 = 0x1b873593;
export const C3 = 0xe6546b64;
export const C4 = 0x85ebca6b;
export const C5 = 0xc2b2ae35;

export const add = (a: number, b: number) => ((a|0) + (b|0)) >>> 0;
export const rot = (a: number, b: number) => ((a << b) | (a >>> (32 - b))) >>> 0;
export const mul = Math.imul;

export const toMurmur53 = (s: string, seed: number = 0) => {
  const n = s.length;
  let a = seed;
  let b = seed ^ C4;

  for (let i = 0; i < n; ++i) {
    let k;
    let d = s.charCodeAt(i);
    let d1 = add(d, b);
    let d2 = add(d, a);

    k = mul(d1, C1);
    k = rot(k, 15);
    k = mul(k, C2);

    a ^= k;
    a = rot(a, 13);
    a = add(mul(a, 5), C3);

    k = mul(d2, C1);
    k = rot(k, 15);
    k = mul(k, C2);

    b ^= k;
    b = rot(b, 13);
    b = add(mul(b, 5), C3);
  }

  a ^= n;
  b ^= n;

  a ^= a >>> 16;
  a = mul(a, C4);
  a ^= a >>> 13;
  a = mul(a, C5);
  a ^= a >>> 16;
  
  b ^= b >>> 16;
  b = mul(b, C4);
  b ^= b >>> 13;
  b = mul(b, C5);
  b ^= b >>> 16;

  return a + ((b & 0x1fffff) * 0x100000000);
}
