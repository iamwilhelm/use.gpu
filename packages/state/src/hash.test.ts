import { getHash } from './hash';

describe('hash', () => {

  it('hashes values', () => {
    expect(getHash(undefined)).toEqual('1l2iguh1n2');
    expect(getHash(null)).toEqual('8vth8a6rwg');
    expect(getHash("")).toEqual('uottwj0y3r');
    expect(getHash("0")).toEqual('38065vocck');
    expect(getHash(0)).toEqual('91bv1y81jg');
    expect(getHash(1)).toEqual('28jo6jjg7z');
    expect(getHash(0x100000000)).toEqual('3l8z18qqcf');
    expect(getHash(0x100000001)).toEqual('bjtjon1r4w');
    expect(getHash([])).toEqual('pcbqk3lwbn');
    expect(getHash([0])).toEqual('yz8mf8hmv4');
    expect(getHash([1])).toEqual('tpn113u1p3');
    expect(getHash({})).toEqual('9q11f1m8s9');
    expect(getHash({f:0})).toEqual('1bru16c4iq');
    expect(getHash({g:0})).toEqual('cd3z2uv6ae');
    expect(getHash({f:1})).toEqual('xd2698tw35');
    expect(getHash({g:1})).toEqual('t80dt2qiwp');

    const u8 = new Uint8Array(3);
    u8[0] = 1;
    u8[1] = 2;

    expect(getHash(u8)).toEqual('zoy66fqire');

    u8[2] = 2;

    expect(getHash(u8)).toEqual('9j8jh6vgmv');
  });
  
});