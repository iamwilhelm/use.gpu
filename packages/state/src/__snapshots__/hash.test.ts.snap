// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hash hashes values 1`] = `
Array [
  "1l2jq1k54d",
  "8vth9qfbt1",
  "uottwj0y3r",
  "38065vocck",
  "ho5ybghprz",
  "3qqy9qztd8",
  "pg9ii5oceu",
  "b5z3tyn70o",
  "pcbqk3lwbn",
  "ko4h5js3v4",
  "3osgp1jl35",
  "mytsog2g27",
  "abikg0mmmd",
  "22mi2di0qp",
  "zoy66fqire",
  "9q11f1m8s9",
  "9a90m9jvkt",
  "74uo9rernc",
  "gn00tx4f37",
  "0hm88fcj85",
]
`;
